packages/block-library/src/navigation/edit/index.js

```diff
@@ -95,13 +95,49 @@ function Navigation( {
 
 	const ref = attributes.ref;
 
-	const setRef = ( postId ) => {
-		setAttributes( { ref: postId } );
+	const setRef = ( postSlug ) => {
+		setAttributes( { ref: postSlug } );
 	};
 
 	const recursionId = `navigationMenu/${ ref }`;
 	const hasAlreadyRendered = useHasRecursion( recursionId );
 
+	const generatedTemplatePartHierarchySlug = useSelect(
+		( select ) => {
+			// Use the lack of a clientId as an opportunity to bypass the rest
+			// of this hook.
+			if ( ! clientId ) {
+				return;
+			}
+
+			const { getBlock, getBlockParentsByBlockName } =
+				select( blockEditorStore );
+
+			const withAscendingResults = true;
+			const parentTemplatePartClientIds = getBlockParentsByBlockName(
+				clientId,
+				'core/template-part',
+				withAscendingResults
+			);
+
+			let result = parentTemplatePartClientIds.map(
+				( templatePartClientId ) => {
+					return getBlock( templatePartClientId )?.attributes?.slug;
+				}
+			);
+
+			// Inclusion of this hardcoded string encodes the template part hierarchy
+			// directly into the generated slug. This may be used later to retrieve
+			// the template part hierarchy from the slug.
+			const TEMPLATE_PART_BOUNDARY_DELINEATOR = '-wptp-';
```

Comment:
Do we need to ever retrieve  the template part hierarchy from the slug? It seems like a thing we should avoid to me.

Reply:
Here's the plan for switching Themes

- you have a block in Theme A which is in the following template hierarchy:

```
- header
-- header-small
--- navigation block
```

- you have a block in Theme B which is in the following template hierarchy:
```
- header
-- navigation block
```
- when you switch Themes the block in Theme B needs to be able to determine which Navigation Menu to show.
- when you created the Navigation Menu in Theme A it will have encoded the template hierarchy at the point of creation within the slug (e.g. `header--header-small`).
- so when switching to Theme B the block can determine it is within the `header` template part which would generated the following slug `header`. Using this it can request the most "appropriate" Navigation Menu from the API by passing the slug.
- in a future PR we'll need to update the REST API endpoint so that it can look up the most "appropriate" Navigation Menu based on the `slug` that is provided in the request.

I've love to hear your thoughts on this and ways where it could be improved our optimised. Perhaps I've made things too complex?

Reply:
I think we should not rely on breaking down the slug unless we can’t otherwise. We make it a central point of failure and it’s just a string.

In my opinion:
- in the editor we look at block hierarchy to determine a place
- we send the hierarchy to the API
- the api makes the slug and looks for similar navigations by post_name
- we then try our best

The slug should only be a thing we look at and say “It looks like this is a good candidate” but to know what **the candidate is for** we look at the data , which is the block hierarchy in the editor.

There is also the problem of theme switching without visiting the editor, what renders on the front end? That means the navigation block should use some WP function to (1) find its hierarchy and (2) render a best guess.

Reply:
Thanks for breaking this down Andrei. I like the idea of sending the template hierarchy data directly rather than encoding that in the slug. However I fear it may fall down in the following scenario...

> There is also the problem of theme switching without visiting the editor, what renders on the front end? That means the navigation block should use some WP function to (1) find its hierarchy and (2) render a best guess.

Exactly. So if the hierarchy is "encoded" within the slug then we can reverse engineer that on the front end to get the most appropriate navigation. 

Without this it's going to be tricky or impossible to reverse engineer the template location unless we save that as a separate attribute which seems redundant.

Reply:
@mtias I'm interested in your perspective here.